pkg/engine/virtualmachines.go

```diff
@@ -34,6 +34,9 @@ func CreateMasterVM(cs *api.ContainerService) VirtualMachineARM {
 	if isStorageAccount {
 		dependencies = append(dependencies, "[variables('masterStorageAccountName')]")
 	}
+	if userAssignedIDEnabled {
+		dependencies = append(dependencies, "[concat('Microsoft.ManagedIdentity/userAssignedIdentities/', variables('userAssignedID'))]")
+	}
```

Comment:
Shall we filter the user assigned identity resource and dependency for scalingup/upgrading template? There is some normalization method in transform.go that filter those infra resources/dependencies like nsg/vnet/rt, etc.

Reply:
Sorry, I'm not familiar with that. What's the purpose of filtering resources and dependencies? To prevent the redeployment swiping existing configuration? If so, for user assigned identity, that will be ok. Even after redeployment, the identity itself, all role assignment on it and all association relationship will keep unchanged; and keep identity in VM's dependency is also fine.

Reply:
Will the identity have the same principal ID after being redeployed?

Reply:
Yes.

Reply:
Ideally when we scale/upgrade a cluster, we should only deploy new VM/VMSS without redeploying the cluster infrastructures, such as VNET/RouteTable/NSG and this UserAssignedIdentity. The consequence of touching these infrastructure is nondeterministic and is tied to how these infra are built/maintained.

 If I were you I would filter them out(Look for these below methods in transform.go)

NormalizeMasterResourcesForScaling
NormalizeForK8sVMASScalingUp
NormalizeResourcesForK8sMasterUpgrade
NormalizeResourcesForK8sAgentUpgrade

Having said that, In this case, the identity is pretty static and if the principal ID doesn't change on redeployment, it seems it's ok if we redeploy it.

Reply:
I think user assigned identity is different from VNET/RouteTable/NSG:
As you said "the consequence of touching these infrastructure is nondeterministic and is tied to how these infra are built/maintained", the biggest difference is it is guaranteed by identity team that a redeployment on a user assigned identity will not change current user assigned identity -- it's a deterministic behavior. And in current AKS-Engine's implementation, the definition of user assigned identity and role assignment on it(at the scope of cluster resource group, that is MC_ resource group in AKS) are not removed from the template when upgrading and scaling, as far as I can see, it brings following benefit:
- When the identity or the role assignment on it is deleted by mistake, a redeployment can bring them back, that is similar as a "reconciliation". And we don't need to worry about a redeployment on user assigned identity will wipe out anything so a redeployment will never harm.

The main purpose of this PR is to add the user assigned identity to VM's `dependsOn`, I think we may want to keep current AKS-Engine's transformer implementation.

Reply:
ping @xuto2